.test-infra/dataproc/init-actions/flink.sh

```diff
@@ -0,0 +1,205 @@
+  # NB: This assumes > 1 worker node.
+  local num_taskmanagers="$(($num_workers - 1))"
+
+  # Determine the number of task slots per worker.
+  # TODO: Dataproc does not currently set the number of worker cores on the
+  # master node. However, the spark configuration sets the number of executors
+  # to be half the number of CPU cores per worker. We use this value to
+  # determine the number of worker cores. Fix this hack when
+  # yarn.nodemanager.resource.cpu-vcores is correctly populated.
+  local spark_executor_cores=$(\
+    grep 'spark\.executor\.cores' /etc/spark/conf/spark-defaults.conf \
+      | tail -n1 \
+      | cut -d'=' -f2)
+  local flink_taskmanager_slots="$(($spark_executor_cores * 2))"
+
+  # Determine the default parallelism.
+  local flink_parallelism=$(python -c \
+    "print ${num_taskmanagers} * ${flink_taskmanager_slots}")
+
+  # Get worker memory from yarn config.
+  local worker_total_mem="$(hdfs getconf \
+    -confKey yarn.nodemanager.resource.memory-mb)"
+  local flink_jobmanager_memory=$(python -c \
+    "print int(${worker_total_mem} * ${FLINK_JOBMANAGER_MEMORY_FRACTION})")
+  local flink_taskmanager_memory=$(python -c \
+    "print int(${worker_total_mem} * ${FLINK_TASKMANAGER_MEMORY_FRACTION})")
+
+  # Fetch the primary master name from metadata.
+  local master_hostname="$(/usr/share/google/get_metadata_value attributes/dataproc-master)"
+
+  # create working directory
+  mkdir -p "${FLINK_WORKING_DIR}"
+
+  # Apply Flink settings by appending them to the default config.
+  cat << EOF >> ${FLINK_INSTALL_DIR}/conf/flink-conf.yaml
+# Settings applied by Cloud Dataproc initialization action
+jobmanager.rpc.address: ${master_hostname}
+jobmanager.heap.mb: ${flink_jobmanager_memory}
+taskmanager.heap.mb: ${flink_taskmanager_memory}
+taskmanager.numberOfTaskSlots: ${flink_taskmanager_slots}
+parallelism.default: ${flink_parallelism}
+taskmanager.network.numberOfBuffers: ${FLINK_NETWORK_NUM_BUFFERS}
+fs.hdfs.hadoopconf: ${HADOOP_CONF_DIR}
+EOF
+
+cat > "${FLINK_YARN_SCRIPT}" << EOF
+#!/bin/bash
+set -exuo pipefail
+sudo -u yarn -i \
+HADOOP_CONF_DIR=${HADOOP_CONF_DIR} \
+  ${FLINK_INSTALL_DIR}/bin/yarn-session.sh \
+  -n "${num_taskmanagers}" \
+  -s "${flink_taskmanager_slots}" \
+  -jm "${flink_jobmanager_memory}" \
+  -tm "${flink_taskmanager_memory}" \
+  -nm flink-dataproc \
+  --detached
+EOF
+chmod +x "${FLINK_YARN_SCRIPT}"
+
+}
+
+function start_flink_master() {
+  local start_yarn_session="$(/usr/share/google/get_metadata_value \
+    "attributes/${START_FLINK_YARN_SESSION_METADATA_KEY}" \
+    || echo "${START_FLINK_YARN_SESSION_DEFAULT}")"
+
+  if ${start_yarn_session} ; then
+    "${FLINK_YARN_SCRIPT}"
+  else
+    echo "Doing nothing"
+  fi
+}
+
+function main() {
+  local role="$(/usr/share/google/get_metadata_value attributes/dataproc-role)"
+  local snapshot_url="$(/usr/share/google/get_metadata_value \
+    "attributes/${FLINK_INSTALL_SNAPSHOT_METADATA_KEY}" \
```

Comment:
Is this populated by dataproc.

Shall we also control the version of flink so that we do not get hit by a flink version change on the dataproc side?

Reply:
Yes - it's populated in the create_flink_cluster.sh in line 128: `metadata+="flink-snapshot-url=${FLINK_DOWNLOAD_URL}`

We have control over flink version. The way it works is that it downloads and installs the snapshot from the url with a version provided in the url.

example url: http://archive.apache.org/dist/flink/flink-1.7.0/flink-1.7.0-bin-hadoop28-scala_2.12.tgz